### ISSUE TITLE
Performance issue with lists of arrays?

### ISSUE THREAD

**Issue by leohaim** [numba/numba#3071]
## Reporting a bug

- [x ] I am using the latest released version of Numba (most recent is visible in
 the change log (https://github.com/numba/numba/blob/master/CHANGE_LOG).
- [x ] I have included below a minimal working reproducer (if you are unsure how
 to write one see http://matthewrocklin.com/blog/work/2018/02/28/minimal-bug-reports).

Hello,
it's great to have lists of numpy arrays now supported in Numba!

I experimented a bit and found that Numba is slower than Numpy for short lists of large arrays but
faster than Numpy for long lists of small arrays. The timings are on a Xeon E5 with Numba 0.39RC1 installed with Conda for python 2.7:

srvx1{~leo/python}165 %python arraylist2.py
short list of large arrays:
Result n=50: 147563.124
Numpy Time: 0.740
Result n=50: 147563.124
Numba Time: 2.120

long list of small arrays:
Result n=500000: 14.441
Numpy Time: 7.186
Result n =500000: 14.441
Numba Time: 1.570

What is the reason for Numba being slower than Numpy in the first example? My use case is more like the first example. 

The reproducer, which is really just for testing:

# -*- coding: utf-8 -*-
"""
Created on Thu Jun 28 18:45:46 2018

@author: Leo
"""

import numpy
from numba import *
import time

def uselist(al):
    lens=numpy.empty(len(al))
    for i in range(len(al)):
        al[i][:,:]*=i
        lens[i]=al[i].shape[1]
        
    return numpy.std(lens)

def genlist(lens):
    al=[]
    for i in range(lens.shape[0]):
        al.append(numpy.zeros((16,lens[i]))+1.0)
    
    return uselist(al)

@njit
def nuselist(al):
    lens=numpy.empty(len(al))
    for i in range(len(al)):
        al[i][:,:]*=i
        lens[i]=al[i].shape[1]
        
    return numpy.std(lens)

@njit
def ngenlist(lens):
    al=[]
    for i in range(lens.shape[0]):
        al.append(numpy.zeros((16,lens[i]))+1.0)
    
    return nuselist(al)

print ('short list of large arrays:')

n=50
m=500000
lens=numpy.random.randint(m,size=n)
al=ngenlist(lens)

for it in range(1):
    t=time.time()
    print('Result n={}: {:5.3f}'.format(n, genlist(lens)))
    print('Numpy Time: {:5.3f}'.format(time.time()-t))

for it in range(1):
    t=time.time()
    print('Result n={}: {:5.3f}'.format(n, ngenlist(lens)))
    print('Numba Time: {:5.3f}'.format(time.time()-t))

print("")
print ('long list of small arrays:')

n=500000
m=50
lens=numpy.random.randint(m,size=n)

for it in range(1):
    t=time.time()
    print('Result n={}: {:5.3f}'.format(n,genlist(lens)))
    print('Numpy Time: {:5.3f}'.format(time.time()-t))

for it in range(1):
    t=time.time()
    print('Result n ={}: {:5.3f}'.format(n,ngenlist(lens)))
    print('Numba Time: {:5.3f}'.format(time.time()-t))

**Comment by stuartarchibald** [comment]
NOTE: Edited report with MD syntax.

**Comment by sklam** [comment]
I think the benchmark is revealing that Numba is lower in allocating large arrays.
Even 

```python

def genlist(lens):
    al = 0
    for i in range(lens.shape[0]):
        al += numpy.sum(numpy.empty(lens[i]))

    return al
```

has a 2x speed difference

**Comment by leohaim** [comment]
I agree. Do you know the reason for this? I thought I would minimize the overhead when creating
the list and its arrays inside the njitted function since the list likely has a more simple representation there.

**Comment by bordingj** [comment]
I am experiencing the same thing. 

Here is another minimal example:


```python
def sum_list(l):
    s = np.zeros_like(l[0])
    for i in range(1,len(l)):
        s += l[i]
    return s

sum_list_nb = nb.njit(sum_list)

l = [np.random.rand(10),]*100

sum_list_nb(l)
```

Gives me

```python
In[33]: %timeit sum_list(l)
71.1 µs ± 72.1 ns per loop (mean ± std. dev. of 7 runs, 10000 loops each)
In[34]: %timeit sum_list_nb(l)
906 µs ± 2.57 µs per loop (mean ± std. dev. of 7 runs, 1000 loops each)
```

**Comment by stuartarchibald** [comment]
This reproduces on Numba 0.51.
